src/Storages/StorageKeeperMap.cpp

```diff
@@ -411,18 +417,13 @@ StorageKeeperMap::StorageKeeperMap(
 
             auto code = client->tryCreate(zk_table_path, "", zkutil::CreateMode::Persistent);
 
-            // tables_path was removed with drop
-            if (code == Coordination::Error::ZNONODE)
-            {
-                LOG_INFO(log, "Metadata nodes were removed by another server, will retry");
-                continue;
-            }
-            else if (code != Coordination::Error::ZOK)
-            {
+            if (code == Coordination::Error::ZOK)
+                return;
```

Comment:
Thus there are two `returns` from this constructor - the first one is here, and the second one is near the end of the constructor. The second return is used when the first replica is added, and the first return is used when any other replica is added, isn't it? It's not really obvious, if we had comments it would be better.

And it seems you set `table_is_valid = true` only for the second return, but not for the first one. Why?

Reply:
I will add comments
but in short, when we add new replicas, we successfully created it but we are still not sure if all nodes are okay and that the metadata matches, so we lazily check the table
the lazy check was added to avoid failing attach if there is no ZK connection

but it would make sense to change it and instantly fail to create a table if the table is invalid, I can do it in another PR